File: apps/iosApp/Sources/ViewModels/ContentDetailViewModel.swift
import Foundation
import SharedKit

enum ContentDetailUiState {
    case loading
    case error(String)
    case loaded(detail: ContentDetail, isInWatchlist: Bool, isUpdatingWatchlist: Bool)
}

@MainActor
final class ContentDetailViewModel: ObservableObject {
    @Published private(set) var uiState: ContentDetailUiState = .loading

    private let contentId: String
    private let contentRepository: ContentRepository
    private let watchlistRepository: WatchlistRepository

    init(contentId: String, contentRepository: ContentRepository, watchlistRepository: WatchlistRepository) {
        self.contentId = contentId
        self.contentRepository = contentRepository
        self.watchlistRepository = watchlistRepository
        load()
    }

    private func load() {
        Task {
            uiState = .loading
            do {
                let detail = try await ThrowingWrappersKt.findByIdOrThrow(contentRepository, contentId: contentId)
                let watchlist = try? await ThrowingWrappersKt.listOrThrow(watchlistRepository, userId: "", limit: 50, cursor: nil)
                let items = (watchlist?.items as? [WatchlistEntry]) ?? []
                let isInList = items.contains { $0.content.id == contentId }
                uiState = .loaded(detail: detail, isInWatchlist: isInList, isUpdatingWatchlist: false)
            } catch {
                uiState = .error(readableMessage(error))
            }
        }
    }

    func toggleWatchlist() {
        guard case .loaded(let detail, let isInWatchlist, _) = uiState else { return }
        uiState = .loaded(detail: detail, isInWatchlist: isInWatchlist, isUpdatingWatchlist: true)
        Task {
            do {
                if isInWatchlist {
                    try await ThrowingWrappersKt.removeOrThrow(watchlistRepository, userId: "", contentId: contentId)
                } else {
                    _ = try await ThrowingWrappersKt.addOrThrow(watchlistRepository, userId: "", contentId: contentId)
                }
                uiState = .loaded(detail: detail, isInWatchlist: !isInWatchlist, isUpdatingWatchlist: false)
            } catch {
                uiState = .loaded(detail: detail, isInWatchlist: isInWatchlist, isUpdatingWatchlist: false)
            }
        }
    }
}

File: apps/androidApp/src/main/kotlin/com/streaming/platform/android/ui/ContentDetailViewModel.kt
package com.streaming.platform.android.ui

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.streaming.platform.content.ContentDetail
import com.streaming.platform.content.ContentRepository
import com.streaming.platform.watchlist.WatchlistRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

sealed interface ContentDetailUiState {
    data object Loading : ContentDetailUiState
    data class Error(val message: String) : ContentDetailUiState
    data class Loaded(val detail: ContentDetail, val isInWatchlist: Boolean, val isUpdatingWatchlist: Boolean = false) : ContentDetailUiState
}

class ContentDetailViewModel(
    private val contentId: String,
    private val contentRepository: ContentRepository,
    private val watchlistRepository: WatchlistRepository,
) : ViewModel() {
    private val _uiState = MutableStateFlow<ContentDetailUiState>(ContentDetailUiState.Loading)
    val uiState: StateFlow<ContentDetailUiState> = _uiState.asStateFlow()

    init { load() }

    private fun load() {
        viewModelScope.launch {
            _uiState.value = ContentDetailUiState.Loading
            val detailResult = contentRepository.findById(contentId)
            val watchlist = watchlistRepository.list("", 50, null).getOrNull()
            detailResult.fold(
                onSuccess = { detail ->
                    val isInList = watchlist?.items?.any { it.content.id == contentId } ?: false
                    _uiState.value = ContentDetailUiState.Loaded(detail, isInList)
                },
                onFailure = { _uiState.value = ContentDetailUiState.Error(it.readableMessage()) },
            )
        }
    }

    fun toggleWatchlist() {
        val current = _uiState.value as? ContentDetailUiState.Loaded ?: return
        _uiState.value = current.copy(isUpdatingWatchlist = true)
        viewModelScope.launch {
            val result = if (current.isInWatchlist) watchlistRepository.remove("", contentId) else watchlistRepository.add("", contentId)
            result.fold(
                onSuccess = { _uiState.value = current.copy(isInWatchlist = !current.isInWatchlist, isUpdatingWatchlist = false) },
                onFailure = { _uiState.value = current.copy(isUpdatingWatchlist = false) },
            )
        }
    }
}
